srml/balances/src/tests.rs

```diff
@@ -21,206 +21,328 @@
+		.existential_deposit(1)
+		.monied(true)
+		.build()
+		.execute_with(|| {
+			Balances::set_lock(
+				ID_1, 
+				&1, 
+				WithdrawLock::Normal(NormalLock {
+					amount: 5, 
+					until: u64::max_value(),
+				}),
+				WithdrawReasons::all()
+			);
+			assert_noop!(
+				<Balances as Currency<_>>::transfer(&1, &2, 6, AllowDeath),
+				"account liquidity restrictions prevent withdrawal"
+			);
+		});
+}
+
+#[test]
+fn lock_reasons_should_work() {
+	ExtBuilder::default()
+		.existential_deposit(1)
+		.monied(true)
+		.build()
+		.execute_with(|| {
+			Balances::set_lock(
+				ID_1, 
+				&1, 
+				WithdrawLock::Normal(NormalLock {
+					amount: 10, 
+					until: u64::max_value(),
+				}),
+				WithdrawReason::Transfer.into()
+			);
+			assert_noop!(
+				<Balances as Currency<_>>::transfer(&1, &2, 1, AllowDeath),
+				"account liquidity restrictions prevent withdrawal"
+			);
+			assert_ok!(<Balances as ReservableCurrency<_>>::reserve(&1, 1));
+			// NOTE: this causes a fee payment.
+			assert!(<ChargeTransactionPayment<Runtime> as SignedExtension>::pre_dispatch(
+				ChargeTransactionPayment::from(1),
+				&1,
+				CALL,
+				info_from_weight(1),
+				0,
+			)
+			.is_ok());
+
+			Balances::set_lock(
+				ID_1, 
+				&1, 
+				WithdrawLock::Normal(NormalLock {
+					amount: 10, 
+					until: u64::max_value(),
+				}),
+				WithdrawReason::Reserve.into()
+			);
+			assert_ok!(<Balances as Currency<_>>::transfer(&1, &2, 1, AllowDeath));
+			assert_noop!(
+				<Balances as ReservableCurrency<_>>::reserve(&1, 1),
+				"account liquidity restrictions prevent withdrawal"
+			);
+			assert!(<ChargeTransactionPayment<Runtime> as SignedExtension>::pre_dispatch(
+				ChargeTransactionPayment::from(1),
+				&1,
+				CALL,
+				info_from_weight(1),
+				0,
+			)
+			.is_ok());
+
+			Balances::set_lock(
+				ID_1,
+				&1,
+				WithdrawLock::Normal(NormalLock {
+					amount: 10,
+					until: u64::max_value(),
+				}),
+				WithdrawReason::TransactionPayment.into(),
+			);
+			assert_ok!(<Balances as Currency<_>>::transfer(&1, &2, 1, AllowDeath));
+			assert_ok!(<Balances as ReservableCurrency<_>>::reserve(&1, 1));
+			assert!(<ChargeTransactionPayment<Runtime> as SignedExtension>::pre_dispatch(
+				ChargeTransactionPayment::from(1),
+				&1,
+				CALL,
+				info_from_weight(1),
+				0,
+			)
+			.is_err());
+		});
+}
+
+#[test]
+fn lock_block_number_should_work() {
+	ExtBuilder::default()
+		.existential_deposit(1)
+		.monied(true)
+		.build()
+		.execute_with(|| {
+			Balances::set_lock(
+				ID_1, 
+				&1, 
+				WithdrawLock::Normal(NormalLock {
+					amount: 10, 
+					until: 2,
+				}),
+				WithdrawReasons::all()
+			);
+			assert_noop!(
+				<Balances as Currency<_>>::transfer(&1, &2, 1, AllowDeath),
+				"account liquidity restrictions prevent withdrawal"
+			);
+
+			Timestamp::set_timestamp(2);
+			assert_ok!(<Balances as Currency<_>>::transfer(&1, &2, 1, AllowDeath));
+		});
+}
+
+#[test]
+fn lock_block_number_extension_should_work() {
```

Comment:
currently we shadow the `LockableCurrency::extend_lock()`

please add TODO to this test case